scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -108,24 +109,346 @@ object ExecutionOptimizationRules {
+            case _ => 0
+          }
+        @annotation.tailrec
+        def loop(fz: FlattenedZip[Execution, Any], acc: Int): Int =
+          fz match {
+            case Single(w) => acc + isW(w)
+            case Many(h, rest, _) => loop(rest, acc + isW(h))
+            case MapZip(of, _) => loop(of, acc)
+          }
+
+        loop(this, 0)
+      }
+
+      /**
+       * The type of this ensures that the head of the HList
+       * must be a WriteExecution. Additionally, none of
+       * the other items are, but we don't prove that in the types,
+       * instead we prove that with tests
+       */
+      def optimize: Option[FlattenedZip[WriteExecution, A]]
+    }
+
+    object FlattenedZip {
+      import HList._
+
+      /**
+       * Here is the simplest FlattenZip: just a single Execution
+       */
+      final case class Single[+Ex[x] <: Execution[x], A](ex: Ex[A]) extends FlattenedZip[Ex, A] {
+        type H = A :: HNil.type
+        type ExH = Execution[A] :: HNil.type
+        def executions = ::(ex, HNil)
+        def toExecution(ex: Execution[A] :: HNil.type): Execution[A :: HNil.type] =
+          ex.head.map(ConsFn(HNil))
+
+        def fn = HeadFn()
+        def zip[B](that: FlattenedZip[Execution, B]) =
+          that match {
+            case s@Single(ex2) => Many[Execution, A, B](ex, s)
+            case m@Many(_, _, _) =>
+              def go[X, Y](m: Many[Execution, X, Y, B]): FlattenedZip[Execution, (A, B)] = {
+                val sub0: SubTypes[(X, Y), B] = m.sub
+                val flat0: FlattenedZip[Execution, (A, (X, Y))] =
+                  Many[Execution, A, (X, Y)](ex, Single(m.head).zip(m.rest))
+                type F[+X] = FlattenedZip[Execution, (A, X)]
+                sub0.liftCo[F](flat0)
+              }
+              go(m)
+            case MapZip(left, fn) =>
+              zip(left).map(ZipMap.MapRight(fn))
+          }
+        def optimize: Option[FlattenedZip[WriteExecution, A]] =
+          ex match {
+            case w@WriteExecution(_, _, _) => Some(Single(w))
+            case _ => None
+          }
+      }
+
+      /**
+       * Here is a non-empty list of Executions. Since scala has a hard time with inferring types
+       * of ADTs we use the Liskov trick here, which we call SubTypes in the scalding codebase.
+       * This allows us to prove without casts when needed that (A, T) <: C which
+       * is something scala has a hard time with.
+       */
+      final case class Many[+Ex[x] <: Execution[x], A, T, +C](head: Ex[A], rest: FlattenedZip[Execution, T], sub: SubTypes[(A, T), C]) extends FlattenedZip[Ex, C] {
+        type H = A :: rest.H
+        type ExH = Execution[A] :: rest.ExH
+        def executions = ::(head, rest.executions)
+        def toExecution(ex: Execution[A] :: rest.ExH): Execution[A :: rest.H] =
+          Mapped(Zipped(ex.head, rest.toExecution(ex.tail)), Cons2Fn())
+
+        def fn: (A :: rest.H) => C = { cons =>
```

Comment:
this is an anonymous function. We should remove this an use a case class function.